### PR TITLE
Allow ORC and Parquet writers to write nullable columns without nulls as non-nullable

### PR DESCRIPTION
**Pull request by vuule**
## Description
Issue #7654, #13010

Writers have a strict check for nullability when applying the user metadata's nullability options, because checking the actual number of nulls is (was) not cheap.
Since we now create all columns with a known number of nulls, the `null_count` check became cheap and we have no reason to prevent columns without nulls to be written as non-nullable.
This PR changes the condition to allow this case. 
The PR does not address the issue with sliced columns, where it's not possible to write sliced column as non-nullable, even if the slice has no nulls. That check is still not cheap :)

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/io/orc/writer_impl.cu [hunk 1]
```diff
@@ -687,9 +687,9 @@ orc_streams create_streams(host_span<orc_column_view> columns,
         // For chunked write, when not provided nullability, we assume the worst case scenario
         // that all columns are nullable.
         auto const chunked_nullable = column.user_defined_nullable().value_or(true);
-        CUDF_EXPECTS(chunked_nullable or !column.nullable(),
-                     "Mismatch in metadata prescribed nullability and input column nullability. "
-                     "Metadata for nullable input column cannot prescribe nullability = false");
+        CUDF_EXPECTS(chunked_nullable or column.null_count() == 0,
```
Comment (ttnghia) [inline cpp/src/io/orc/writer_impl.cu:690]: Probably you also need to update the comment above it.

Comment (vuule) [reply cpp/src/io/orc/writer_impl.cu:690]: The comment refers to the unchanged line above, tries to explain the default `true` value.

File: cpp/tests/io/orc_test.cpp [hunk 2]
```diff
@@ -1806,4 +1806,21 @@ TEST_F(OrcWriterTest, EmptyRowGroup)
...
+
+  cudf::io::table_input_metadata expected_metadata(expected);
+  expected_metadata.column_metadata[0].set_nullability(false);
+
+  auto filepath = temp_env->get_temp_filepath("NonNullable.orc");
+  cudf::io::orc_writer_options out_opts =
+    cudf::io::orc_writer_options::builder(cudf::io::sink_info{filepath}, expected)
+      .metadata(&expected_metadata);
+  // Writer should be able to write a column without nulls as non-nullable
+  EXPECT_NO_THROW(cudf::io::write_orc(out_opts));
```
Comment (harrism) [inline cpp/tests/io/orc_test.cpp:1823]: Just a nit, either way is fine, but I don't think EXPECT_NO_THROW is required. If it throws and the exception is not caught, the test fails anyway.

**Review by harrism — APPROVED** [review]

File: cpp/tests/io/orc_test.cpp (hunk 2 above)
Comment (vuule) [reply cpp/tests/io/orc_test.cpp:1823]: I use it just to make it easier to see what we're checking. I prefer to have the macro for this reason.

**Review by ttnghia — APPROVED** [review]

**Pull request merged**